### ISSUE TITLE
C Exceptions aren't translated to Python Exceptions, thusly crashing Python

### ISSUE THREAD

**Issue by alexander-gabriel** [opencog/atomspace#2518]
Here is an example that results in such a core dump:

```
#!/usr/bin/env python

from opencog.type_constructors import *
from opencog.bindlink import execute_atom, evaluate_atom
from opencog.ure import BackwardChainer, ForwardChainer
from opencog.utilities import initialize_opencog
from opencog.atomspace import AtomSpace, types, TruthValue, create_child_atomspace
from opencog.logger import create_logger, Logger,log
from opencog.scheme_wrapper import scheme_eval

TRUE = TruthValue(1.0, 1.0)

def identical_formula(A, B):
    # rospy.loginfo("identical_formula({},{})".format(A,B))
    link = IdenticalLink(A,B)
    # out = A.get_out()
    if A == B:
        link.tv = TruthValue(1, 1)
    else:
        link.tv = TruthValue(0, 1)
    return link


def build_deduction_rulebase():
    rbs = ConceptNode("deduction-rule-base")
    execute_code = '''
    
    (use-modules (opencog))
    (use-modules (opencog logger) (opencog ure) (opencog exec))
    (load-from-path "/home/rasberry/git/pln/opencog/pln/rules/propositional/crisp-conjunction-introduction.scm")
    (load-from-path "/home/rasberry/git/pln/opencog/pln/rules/propositional/fuzzy-conjunction-introduction.scm")

    (load-from-path "/home/rasberry/git/pln/opencog/pln/rules/wip/negation-introduction.scm")

    (define rbs (Concept "deduction-rule-base"))
    (ure-set-complexity-penalty rbs 0.1)
    '''
    scheme_eval(atomspace, execute_code)
    MemberLink(DefinedSchemaNode("fuzzy-conjunction-introduction-2ary-rule"), rbs)
    MemberLink(DefinedSchemaNode("fuzzy-conjunction-introduction-3ary-rule"), rbs)
    MemberLink(DefinedSchemaNode("fuzzy-conjunction-introduction-4ary-rule"), rbs)
    MemberLink(DefinedSchemaNode("fuzzy-conjunction-introduction-5ary-rule"), rbs)
    MemberLink(DefinedSchemaNode("negation-introduction-rule"), rbs)
    return rbs

def print_results(query,results):
    print("Query:\n{:}\n\nResults:\n{:}\n\nDetails:\n--------------------------".format(query, results))
    if query.type_name == "GetLink":
        for setlink in results.get_out():
            for result in setlink.get_out():
                print("Result Truth: {:}".format(result.tv))
                print("Result:\n{:}".format(result))
                print("------------------------\n")
    elif query.type_name == "AndLink":
        for result in results.get_out():
            print("Result Truth: {:}".format(result.tv))
            print("Result:\n{:}".format(result))
            for condition in result.get_out():
                print("------------------------")
                print("Condition:{:}".format(condition))
                print("Condition Truth: {:}".format(condition.tv))
                if condition.type_name == "NotLink":
                    subcondition = condition.get_out()[0]
                    print("    Subcondition: {:}".format(subcondition))
                    print("    Subcondition Truth: {:}".format(subcondition.tv))
            print("-----------------------------------------------------")
    else:
        for result in results.get_out():
            print("Result Truth: {:}".format(result.tv))
            print("Result:\n{:}".format(result))
            print("------------------------\n")


if __name__ == "__main__":
    atomspace = AtomSpace()
    # logger = Logger()
    logger = log
    # logger.set_component("URE")
    # logger.set_level("DEBUG")
    # logger.use_stdout()
    initialize_opencog(atomspace)
    rbs = build_deduction_rulebase()
    true = ConceptNode("true")
    false = ConceptNode("false")
    place1 = ConceptNode("place1")
    place2 = ConceptNode("place2")
    place3 = ConceptNode("place3")
    human = ConceptNode("human").truth_value(1.0, 1.0)
    alice = ConceptNode("alice").truth_value(1.0, 1.0)
    bob = ConceptNode("bob").truth_value(1.0, 1.0)

    StateLink(alice, place1).truth_value(1.0, 1.0)
    StateLink(bob, place2).truth_value(1.0, 1.0)

    EvaluationLink(
        PredicateNode("leads_to"),
        ListLink(place1, place2))

    EvaluationLink(
        PredicateNode("leads_to"),
        ListLink(place1, place3))

    InheritanceLink(alice, human).truth_value(1.0, 1.0)
    InheritanceLink(bob, human).truth_value(1.0, 1.0)

    variables = VariableList(
        TypedVariableLink(VariableNode("person1"), TypeNode("ConceptNode")),
        TypedVariableLink(VariableNode("origin"), TypeNode("ConceptNode")),
        TypedVariableLink(VariableNode("destination"), TypeNode("ConceptNode")),
        )

    query =     AndLink(
                    InheritanceLink(VariableNode("person1"), human),
                    StateLink(VariableNode("person1"), VariableNode("origin")),
                    EvaluationLink(
                        PredicateNode("leads_to"),
                        ListLink(VariableNode("origin"), VariableNode("destination"))
                    ),
                    AbsentLink(
                            StateLink(VariableNode("person2"), VariableNode("destination")))
                    )

    query = GetLink(variables, query)
    trace = create_child_atomspace(atomspace)
    chainer = BackwardChainer(_as=atomspace,
                      rbs=rbs, trace_as=trace, control_as=None, focus_set=None,
                      target=query, vardecl=variables)
    chainer.do_chain()

    results = chainer.get_results()
    print_results(query, results)
```

**Comment by linas** [comment]
@vsbogd I'm throwing this at you, as a python maintainer, but maybe @noskill would be interested in taking a look. The above is a very complicated example; I suspect that even simple examples that trigger C++ exceptions are probably broken...

**Comment by linas** [comment]
So I am looking at the source code.  I'm guessing that this line: 

https://github.com/opencog/atomspace/blob/390ab1713018148b0c18b6a0c70f08b292358eee/opencog/cython/PythonEval.cc#L1184

should have been wrapped in a try..catch to catch C++ errors. And also this: 

https://github.com/opencog/atomspace/blob/390ab1713018148b0c18b6a0c70f08b292358eee/opencog/cython/PythonEval.cc#L1317-L1319

and also this:

https://github.com/opencog/atomspace/blob/390ab1713018148b0c18b6a0c70f08b292358eee/opencog/cython/PythonEval.cc#L966

I'm pretty sure that, in the past, these used to be wrapped in try..catch so I'm confused as to how they got unwrapped... I'm guessing there there is no unit test, or the unit test got lost?

**Comment by linas** [comment]
Actually, the above comments might be totally wrong. For example, I see this:

https://github.com/opencog/atomspace/blob/390ab1713018148b0c18b6a0c70f08b292358eee/opencog/cython/PythonEval.cc#L943-L944

and this

https://github.com/opencog/atomspace/blob/390ab1713018148b0c18b6a0c70f08b292358eee/opencog/cython/PythonEval.cc#L983

and sixteen(!) other place where that code is throwing C++ exceptions. So who is catching these exceptions?  If I `grep -r try *` I can see three `pyx` files that catch python exceptions, but none that would handle these.  Where in the world is the exception-handling code?

**Comment by linas** [comment]
Hmmm.. both of my comments above are wrong, because this simple example does work:
```
>>> from opencog.atomspace import AtomSpace
>>> from opencog.atomspace import types
>>> from opencog.type_constructors import *
>>> a = AtomSpace()
>>> set_type_ctor_atomspace(a)
>>> A = ConceptNode('Apple')
>>> B = GetLink(A, A, A)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "core_types.pyx", line 115, in opencog.type_constructors.GetLink
  File "utilities.pyx", line 67, in opencog.utilities.add_link
RuntimeError: Expecting (optional) variable decls and a body; got (GetLink
  (ConceptNode "Apple") ; [158ed8540a84fda2][1]
  (ConceptNode "Apple") ; [158ed8540a84fda2][1]
  (ConceptNode "Apple") ; [158ed8540a84fda2][1]
) ; [ad9aa4995405eeb1][-1] (/home/linas/src/novamente/src/atomspace-git/opencog/atoms/pattern/PatternLink.cc:154)
```
so for simple examples, the C++ exceptions are being handled correctly. I have no clue who is catching those exceptions, or why the above more complex example is failing.  Oh well... I tried.

**Comment by linas** [comment]
OK, so here is a simple example that does fail:
```
>>> from opencog.atomspace import AtomSpace, TruthValue
>>> from opencog.atomspace import types
>>> from opencog.type_constructors import *
>>> a = AtomSpace()
>>> set_type_ctor_atomspace(a)
>>> A = ConceptNode('Apple')
>>> P = GroundedPredicateNode('py:foobar')
>>> E = EvaluationLink(P,A,A)
>>> from opencog.bindlink import evaluate_atom
>>> evaluate_atom(a, E)
terminate called after throwing an instance of 'opencog::RuntimeException'
  what():  Python function 'foobar' not found in module '__main__'! (/home/linas/src/novamente/src/atomspace-git/opencog/cython/PythonEval.cc:920)
Aborted
```

so it looks like any exception thrown by `PythonEval.cc` is not caught... 

The exceptions that are caught are because of this:

https://github.com/opencog/atomspace/blob/390ab1713018148b0c18b6a0c70f08b292358eee/opencog/cython/opencog/utilities.pxd#L12

**Comment by linas** [comment]
OK so #2519 fixes my last example, but the first example still crashes:
```
ERROR: In procedure apply-smob/1:
In procedure primitive-load-path: Unable to find file "/home/rasberry/git/pln/opencog/pln/rules/propositional/crisp-conjunction-introduction.scm" in load path
ABORT: misc-error
 (/home/linas/src/novamente/src/atomspace-git/opencog/cython/opencog/PyScheme.cc:63)
```
even though `eval_scheme` has an `except +` on it...

... and the original but report also still fails...

**Comment by linas** [comment]
This is fixed in opencog/ure#68

**Comment by linas** [comment]
Closing; I believe this is fixed (or will be when  opencog/ure#68 is merged) please open new bug if problems persist.

**Comment by linas** [comment]
BTW AbsentLink now works, when #2521 is merged.
